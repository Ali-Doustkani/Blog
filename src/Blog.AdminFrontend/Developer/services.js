async function getDeveloper() {
   const response = await fetch('/api/developer')
   return await response.json()
}

async function saveDeveloper(developer) {
   const response = await fetch('/api/developer', {
      method: 'POST',
      headers: {
         'Content-Type': 'application/json'
      },
      body: JSON.stringify(developer)
   })
   return await response.json()
}

export { getDeveloper, saveDeveloper }
