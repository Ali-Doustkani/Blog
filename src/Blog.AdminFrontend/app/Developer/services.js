const url_developer =
   process.env.NODE_ENV === 'production' ? '/api/developer' : 'http://localhost:3000/api/developer'

async function getDeveloper() {
   try {
      const response = await fetch(url_developer, {
         headers: {
            Accept: 'application/json'
         }
      })
      if (response.status === 200) {
         const developer = await response.json()
         return {
            status: 'ok',
            developer
         }
      }
      if (response.status === 204) {
         return {
            status: 'ok',
            developer: null
         }
      }
      if (response.status === 404) {
         return {
            status: 'error',
            message: 'server not found'
         }
      }
      return {
         status: 'error',
         message: 'problem loading data'
      }
   } catch (err) {
      return {
         status: 'error',
         message: err.message
      }
   }
}

async function saveDeveloper(developer) {
   try {
      const response = await fetch(url_developer, {
         method: 'PUT',
         headers: {
            'Content-Type': 'application/json'
         },
         body: JSON.stringify(developer)
      })
      if (response.status === 200) {
         const data = await response.json()
         return {
            status: 'ok',
            data
         }
      }
   } catch (ex) {
      return {
         status: 'error',
         message: ex.message
      }
   }
}

export { getDeveloper, saveDeveloper }
