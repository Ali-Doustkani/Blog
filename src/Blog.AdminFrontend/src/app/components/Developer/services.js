import { emptyValidator, richtextEmptyValidator } from '../../utils'

const url_developer =
   process.env.NODE_ENV === 'production' ? '/api/developer' : 'http://localhost:3000/api/developer'

const ok = data => ({
   status: 'ok',
   data: data || null
})

const error = data => ({
   status: 'error',
   data
})

const fatal = msg => ({
   status: 'fatal',
   data: msg
})

const processResponse = async response => {
   switch (response.status) {
      case 200:
      case 201:
         return ok(await response.json())
      case 204:
         return ok()
      case 400:
         return error(await response.json())
      case 404:
         return fatal('server not found')
      default:
         return fatal(await response.text())
   }
}

const getDeveloper = async () => {
   try {
      return await processResponse(
         await fetch(url_developer, {
            headers: {
               Accept: 'application/json'
            }
         })
      )
   } catch (err) {
      return fatal(err.message)
   }
}

const saveDeveloper = async state => {
   try {
      return await processResponse(
         await fetch(url_developer, {
            method: 'PUT',
            headers: {
               'Content-Type': 'application/json'
            },
            body: JSON.stringify(toDeveloper(state))
         })
      )
   } catch (err) {
      return fatal(err.message)
   }
}

const toDeveloper = state => {
   const experiences = state.experiences.map(exp => ({
      id: exp.id,
      company: exp.company,
      position: exp.position,
      startDate: exp.startDate,
      endDate: exp.endDate,
      content: exp.content
   }))
   return {
      summary: state.summary,
      skills: state.skills,
      experiences
   }
}

const clientError = err => err.type === 1

const anyError = state =>
   Boolean(
      state.summaryErrors.some(clientError) ||
         state.skillsErrors.some(clientError) ||
         state.experiencesErrors.some(clientError) ||
         state.experiences.some(
            exp =>
               exp.companyErrors.some(clientError) ||
               exp.positionErrors.some(clientError) ||
               exp.startDateErrors.some(clientError) ||
               exp.endDateErrors.some(clientError) ||
               exp.contentErrors.some(clientError)
         )
   )

const validate = state => {
   state.summaryErrors = richtextEmptyValidator('summary', state.summaryErrors)(state.summary)
   state.skillsErrors = emptyValidator('skills', state.skillsErrors)(state.skills)
   state.experiences.forEach(validateExperience)
}

const validateExperience = experience => {
   experience.companyErrors = emptyValidator('company', experience.companyErrors)(
      experience.company
   )
   experience.positionErrors = emptyValidator('position', experience.positionErrors)(
      experience.position
   )
   experience.startDateErrors = emptyValidator('start date', experience.startDateErrors)(
      experience.startDate
   )
   experience.endDateErrors = emptyValidator('end date', experience.endDateErrors)(
      experience.endDate
   )
   experience.contentErrors = richtextEmptyValidator('content', experience.contentErrors)(
      experience.content
   )
   return experience
}

export { getDeveloper, saveDeveloper, anyError, validate }
