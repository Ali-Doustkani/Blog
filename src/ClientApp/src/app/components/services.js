const url_developer = '/api/developer'
const url_posts = '/api/posts'

const ok = data => ({
   status: 'ok',
   data: data || null
})

const error = data => ({
   status: 'error',
   data
})

const fatal = msg => ({
   status: 'fatal',
   data: msg
})

const processResponse = async response => {
   switch (response.status) {
      case 200:
      case 201:
         return ok(await response.json())
      case 204:
         return ok()
      case 400:
         return error(await response.json())
      case 401:
         return fatal('not authenticated')
      case 403:
         return fatal('forbidden')
      case 404:
         return fatal('server not found')
      default:
         return fatal(await response.text())
   }
}

const getDeveloper = async auth0 => {
   try {
      const token = await auth0.getAccessToken()
      return await processResponse(
         await fetch(url_developer, {
            headers: {
               Accept: 'application/json',
               Authorization: 'Bearer ' + token
            }
         })
      )
   } catch (err) {
      return fatal(err.message)
   }
}

const saveDeveloper = async (state, auth0) => {
   try {
      const token = await auth0.getAccessToken()
      return await processResponse(
         await fetch(url_developer, {
            method: 'PUT',
            headers: {
               'Content-Type': 'application/json',
               Authorization: 'Bearer ' + token
            },
            body: JSON.stringify(state)
         })
      )
   } catch (err) {
      return fatal(err.message)
   }
}

const getPostItems = async () => {
   try {
      return await processResponse(
         await fetch(url_posts, {
            headers: {
               Accept: 'application/json'
            }
         })
      )
   } catch (err) {
      return fatal(err.message)
   }
}

export { getDeveloper, saveDeveloper, getPostItems }
