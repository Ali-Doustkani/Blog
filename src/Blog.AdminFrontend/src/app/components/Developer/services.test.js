import fetchMock from 'fetchMock'
import { getDeveloper, saveDeveloper } from './services'

jest.mock('../../utils')

const sample = {
   summary: 'started to code as a young boy',
   skills: 'c#, javascript',
   experiences: []
}

beforeEach(fetchMock.reset)

describe('GET', () => {
   it('handles 200', async () => {
      fetchMock.status(200).data({ name: 'developer' })

      expect(await getDeveloper()).toEqual({
         status: 'ok',
         data: { name: 'developer' }
      })
   })

   it('handles 204', async () => {
      fetchMock.status(204)

      expect(await getDeveloper()).toEqual({
         status: 'ok',
         data: null
      })
   })

   it('handles 404', async () => {
      fetchMock.status(404)

      expect(await getDeveloper()).toEqual({
         status: 'fatal',
         data: 'server not found'
      })
   })

   it('handles other status', async () => {
      fetchMock.status(500).data('error message')

      expect(await getDeveloper()).toEqual({
         status: 'fatal',
         data: 'error message'
      })
   })

   it('handles corrupted data', async () => {
      fetchMock.status(200).throwOnData('corrupted data')

      expect(await getDeveloper()).toEqual({
         status: 'fatal',
         data: 'corrupted data'
      })
   })

   it('handles ERR_CONNECTION_REFUSED', async () => {
      fetchMock.throw('Error')

      expect(await getDeveloper()).toEqual({
         status: 'fatal',
         data: 'Error'
      })
   })
})

describe('PUT', () => {
   it('handles 200', async () => {
      fetchMock.status(200).data({ data: 'response' })

      expect(await saveDeveloper(sample)).toEqual({
         status: 'ok',
         data: { data: 'response' }
      })
      expect(fetchMock.body).toEqual(JSON.stringify(sample))
   })

   it('handles 200 with corrupted content', async () => {
      fetchMock.status(200).throwOnData('corrupted data')

      expect(await saveDeveloper(sample)).toEqual({
         status: 'fatal',
         data: 'corrupted data'
      })
   })

   it('handles 201', async () => {
      fetchMock.status(201).data({ data: 'result' })

      expect(await saveDeveloper(sample)).toEqual({
         status: 'ok',
         data: { data: 'result' }
      })
   })

   it('handles 400', async () => {
      fetchMock.status(400).data({
         summary: ['summary field is required']
      })

      expect(await saveDeveloper(sample)).toEqual({
         status: 'error',
         data: { summary: ['summary field is required'] }
      })
   })

   it('handles 404', async () => {
      fetchMock.status(404)

      expect(await saveDeveloper(sample)).toEqual({
         status: 'fatal',
         data: 'server not found'
      })
   })

   it('handles other statuses', async () => {
      fetchMock.status(500).data('error on server')

      expect(await saveDeveloper(sample)).toEqual({
         status: 'fatal',
         data: 'error on server'
      })
   })

   it('handle ERR_CONNECTION_REFUSED', async () => {
      fetchMock.throw('ERROR')

      expect(await saveDeveloper(sample)).toEqual({
         status: 'fatal',
         data: 'ERROR'
      })
   })
})
