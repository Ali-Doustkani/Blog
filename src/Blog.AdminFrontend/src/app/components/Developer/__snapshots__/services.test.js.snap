// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validates company fields 1`] = `
Object {
  "experiences": Array [
    Object {
      "company": "company",
      "companyErrors": "companyError",
      "content": "content",
      "contentErrors": "contentError",
      "endDate": "endDate",
      "endDateErrors": "end dateError",
      "position": "position",
      "positionErrors": "positionError",
      "startDate": "startDate",
      "startDateErrors": "start dateError",
    },
  ],
  "skills": "skills",
  "skillsErrors": "skillsError",
  "summary": "summary",
  "summaryErrors": "summaryError",
}
`;
